### ISSUE TITLE
Discuss desired features/syntax for persistence convenience js/py API

### ISSUE THREAD

**Issue by bollwyvl** [pyodide/pyodide#1715]
As a follow-on to #1596, which exposes the emscripten [FS](https://emscripten.org/docs/api_reference/Filesystem-API.html#persistent-data) directly, more useful wrapper could be added to the persistence API.

Some ideas:

- on the JS side, add additional options to `loadPyodide` to auto-mount a backend
  - the tokens (e.g. `pyodide.FS.filesystems.IDBFS`) would not yet be avaialble, so it might have to be by name
  - it might be unclear how to persist data back out, but a naive polling solution would work
- on the python side, offer a some context managers, etc. that would wrap up the lifecycle of:
  - `mount`, `syncfs(true)`, _do some stuff_, `syncfs(false)`

**Comment by bollwyvl** [comment]
For example, rather than node-style callbacks, this (reverted) commit offered a [promise wrapper](https://github.com/pyodide/pyodide/pull/1596/commits/ddc8421bb55b2c9be794942643bb11b569ba5cce
). This sorta lives on in the [tests](https://github.com/pyodide/pyodide/pull/1596/commits/86f6856cc88951a79101d440275ba7afd548cd7e).

**Comment by hoodmane** [comment]
Promise wrappers would be great.

**Comment by xhlulu** [comment]
I'm not familiar with emscripten, but I saw the tweet about https://github.com/pyodide/pyodide/pull/1596 and I found the elevator pitch and motivation to be quite interesting. Came here to ask: will the new 0.18 release enable an "easy" way to statically served files into the virtual python FS? Currently [I'm using base64 encoding/decoding](https://github.com/xhlulu/pyodide-multi-module-template) (see https://github.com/pyodide/pyodide/issues/419#issuecomment-881760931), but curious if this might be a cleaner alternative.

**Comment by hoodmane** [comment]
Yeah this should be a better alternative. In particular to make a file from some url you can do something like:
```js
let data = await (await fetch(some_url)).text();
pyodide.FS.writeFile(some_path, data);
```

**Comment by bollwyvl** [comment]
I've found that using `.text` can sometimes be troublesome in different browsers... i stole the `arrayBuffer` trick from [micropip](https://github.com/pyodide/pyodide/blob/0.18.0/packages/micropip/src/micropip/micropip.py#L60) for some large-ish stuff.

While we _are_ up (but not released) on 0.18 over on jupyterlite, I haven't yet been able to find a pattern with the existing FS implementations that i like, though I do have some thoughts:

- `PROXYFS` (which is all but undocumented) might be nice in our case, as we could share filesystems between multiple kernels
- [browserfs](https://github.com/jvilk/BrowserFS/blob/a96aa2d417995dac7d376987839bc4e95e218e06/src/generic/emscripten_fs.ts#L346) shows a full implementation of the required implementation of a novel filesystem, if you've already got one
  - I'm kinda leaning towards writing one of these which wraps the JupyterLite file system, which is itself an overlay of some other (eventually pluggable) things)

**Comment by stefnotch** [comment]
> * [browserfs](https://github.com/jvilk/BrowserFS/blob/a96aa2d417995dac7d376987839bc4e95e218e06/src/generic/emscripten_fs.ts#L346) shows a full implementation of the required implementation of a novel filesystem, if you've already got one
>   * I'm kinda leaning towards writing one of these which wraps the JupyterLite file system, which is itself an overlay of some other (eventually pluggable) things)

The browserfs implementation appears to mostly be a typesafe variant of an older version of https://github.com/emscripten-core/emscripten/blob/main/src/library_nodefs.js

Regarding writing one of these, it should be quite doable. May I suggest writing a very generic one of these, which by default implements as many methods as possible using other methods (e.g. rename is "read file, delete file, create new file with content"). I'm also currently in the process of checking out the filesystems and I'm heavily leaning towards writing my own for [starboard](https://github.com/gzuidhof/starboard-notebook/). If we could join forces, that'd be amazing!

**Comment by bollwyvl** [comment]
> a very generic one

While writing a minimal/generic one is attractive, we've already dug ourselves a fairly deep hole on JupyterLite, and implement the actual Jupyter Contents Model (warts and all) as that's what all the extensions expect. I don't know how much would be directly shareable outside of that context, but it will at least have a semi-generic TS interface (part of why i was interested in `browserfs`, which I would use directly, if it appeared maintained). 

>  as many methods as possible using other methods 

At present, we (semi-successfully) do an overlay filesystem of `localForage` and for-real served HTTP files with pre-calculated metadata, and want to be able to support Big Media (videos, large datasets, etc.). So we will need to be a bit ginger with some of those kinds of things.

**Comment by stefnotch** [comment]
@bollwyvl I see, that makes sense. In that case, one thing that we might be able to share would be Typescript type definitions and documentation, as Emscripten has none of that.

e.g. It would be rather useful to know what this function means and what types the parameters have
https://github.com/emscripten-core/emscripten/blob/7c8be3dab1218d74c0081636e2321febdacda43f/src/library_nodefs.js#L306


Also, some good news is that the emscripten maintainers are interested in improving the filesystem code.
https://github.com/emscripten-core/emscripten/issues/9372#issuecomment-898739048

**Comment by stefnotch** [comment]
Quick update: I wrote some code and it seems to work:

1. A [generic filesystem interface](https://github.com/gzuidhof/starboard-notebook/pull/58/files#diff-af99d904915beb79b01bf148c8c1c6f2a5b270fc1bdceb7ae812a86f314c6d1dR14-R41). Notice how this one is promise based.
2. [All the code that translates between emscripten and that generic filesystem interface.](https://github.com/gzuidhof/starboard-python/commit/0fb45fd3773802c34023eecab4912b0b5822c2c3) That involves some [synchronous IO](https://github.com/pyodide/pyodide/issues/1503) code, which has the convenient advantage of letting me use emscripten's synchronous filesystem API

**Comment by stefnotch** [comment]
I worked a bit more on the filesystem code and documented some [issues I ran into here](https://github.com/emscripten-core/emscripten/issues/9372#issuecomment-900118108). 

Now that I've mostly sorted them out, I can say that the filesystem works pretty well. Here is a screenshot of Pyodide in a web worker talking to a filesystem on the main thread.

![image](https://user-images.githubusercontent.com/10220080/129696327-c58b37a1-6b85-4718-8d9c-45aba59f249b.png)

For reference, here is the filesystem Javascript code https://github.com/gzuidhof/starboard-python/issues/9
The idea is to have a very simple API that can be implemented by pretty much any filesystem. This very much includes the [Jupyter Contents REST API](https://jupyter-server.readthedocs.io/en/latest/developers/rest-api.html)
